### PR TITLE
Ignore extra query parameters from referer url when checking for the last visited category

### LINKED ISSUE #20946: Wrong current category on the product page when browsing paginated product list
#### Describe the bug

When a category has many products, the category page lists these products with a pagination.
Clicking on a product from the second or more page will change the "current" category to the product's default instead of the category being browsed.

#### Expected behavior

The "current" category should be the last visited one

1. Go to BO > Design > Position > Transplant the `ps_categorytree` module to `displayProductAdditionalInfo` hook
2. Go to BO > Products > edit the demo product id_product=7 with default category = Home & edit the demo product id_product=8 with default category = Home
image
3. Go to BO > Shop Parameters > Product Settings > set Product per page =2
4. Go to http://shop.com/en/6-accessories
5. Choose the product "MUG THE ADVENTURE BEGINS"
-> Notice the breadcrumb, the default category is Home
-> Notice the ps_categorytree under it, the current category is "Accessories"
6. Go to http://shop.com/en/6-accessories
7. Click on the "Next" in the pagination
8. Choose the product "MUG TODAY IS A GOOD DAY"
-> Notice the breadcrumb,  the default category is also Home
-> Notice the ps_categorytree under it, current category is "Home"

1. Have a category of products "T-Shirts"
2. Create many products that belong to the "T-Shirts" category and which default category is "Home"
3. Got to the "T-Shirts" category
4. Click on a product -> the current category is "T-Shirts" (<strong>This is correct</strong>)
5. Go to the second page of pagination of the "T-Shirts" category
6. Click on a product -> the current category is "Home" (<strong>This is NOT correct</strong>)

This affects in particular the behaviour of the `ps_categorytree` module.

**Screenshots**

#### Additional information

PR to follow

### PR DESCRIPTION
**Pull request by jbenezech**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Strip query parameters when matching referer against category links or pagination parameters will break the match
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #20946
| How to test?  | See issue description

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20947)

### CODE REVIEW

**Review by sowbiba — DISMISSED** [review]

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
With this kind of thing you force the user to work with URL rewriting. Since we allow both (with or without) referees without URL rewrite are broken :thinking:

**Comment by jbenezech** [comment]
Well spotted @PierreRambaud (my first Prestashop project so not too familar with it yet ;)
I actually did not even know url rewrite could be disabled...

What would be a better solution you think ?

Change
`if (in_array($this->context->link->getCategoryLink($id_object), $referers)) {`

to check each referer "begins with"  the category link ? Would that be consistent ?

File: controllers/front/ProductController.php [hunk 1]
```diff
@@ -1323,4 +1312,65 @@ protected function isValidCombination($productAttributeId, $productId)
...
+
+        return null;
+    }
+
+    /**
+     * Given an id and a page link, extracts the category id
+     *
+     * @return int|null
+     */
+    private function extractObjectCategoryFromLink($id_object, $previousPageLink)
```
Comment (Progi1984) [inline controllers/front/ProductController.php]:
```suggestion
    private function extractObjectCategoryFromLink($id_object, $previousPageLink): ?int
```

File: controllers/front/ProductController.php [hunk 2]
```diff
@@ -1323,4 +1312,65 @@ protected function isValidCombination($productAttributeId, $productId)
 
         return false;
     }
+
+    /**
+     * From the current request, tries to determine the current category being browsed
+     *
+     * @return int|null
+     */
+    protected function guessCategoryIdFromRequest()
```
Comment (Progi1984) [inline controllers/front/ProductController.php]:
```suggestion
    protected function guessCategoryIdFromRequest(): ?int
```

File: controllers/front/ProductController.php [hunk 3]
```diff
@@ -1323,4 +1312,65 @@ protected function isValidCombination($productAttributeId, $productId)
...
+        }
+
+        if (!$id_object) {
+            return null;
+        }
+
+        foreach ([$_SERVER['HTTP_REFERER'], urldecode($_SERVER['HTTP_REFERER'])] as $previousPageLink) {
+            $id_category = $this->extractObjectCategoryFromLink($id_object, $previousPageLink);
+            if ($id_category !== null) {
+                return $id_category;
```
Comment (Progi1984) [inline controllers/front/ProductController.php:1394]:
```suggestion
                return (int) $id_category;
```

**Review by Progi1984 — COMMENTED** [review]
Small feedbacks but good job. The code is more readable.

File: controllers/front/ProductController.php (hunk 3 above)
Comment (jbenezech) [reply controllers/front/ProductController.php:1394]: No harm but seems a bit unnecessary if the `extractObjectCategoryFromLink` is typed hinted no ?

Comment (Progi1984) [reply controllers/front/ProductController.php:1394]: @jbenezech Correct. I close this one. (Thanks ;) )

**Review by Progi1984 — DISMISSED** [review]

File: controllers/front/ProductController.php [hunk 4]
```diff
@@ -1323,4 +1312,65 @@ protected function isValidCombination($productAttributeId, $productId)
...
+    private function extractObjectCategoryFromLink($id_object, $previousPageLink): ?int
+    {
+        if (strncmp($this->context->link->getCategoryLink($id_object), $previousPageLink, strlen($this->context->link->getCategoryLink($id_object))) === 0) {
+            return (int) $id_object;
+        }
+
+        if (
+            isset($this->context->cookie->last_visited_category) &&
+            (int) $this->context->cookie->last_visited_category &&
+            strncmp($this->context->link->getProductLink($id_object), $previousPageLink, strlen($this->context->link->getCategoryLink($id_object))) === 0
```
Comment (Progi1984) [inline controllers/front/ProductController.php]:
```suggestion
            strncmp($this->context->link->getProductLink($id_object), $previousPageLink, strlen($this->context->link->getProductLink($id_object))) === 0
```

?

**Review by Progi1984 — COMMENTED** [review]
A small question

**Comment by PierreRambaud** [comment]
@PrestaShop/prestashop-core-developers 
What is the status of this PR?

File: controllers/front/ProductController.php [hunk 5]
```diff
@@ -1369,4 +1358,65 @@ public function setQuickViewMode(bool $enabled = true)
...
+     * @return int|null
+     */
+    protected function guessCategoryIdFromRequest(): ?int
+    {
+        // Assure us the previous page was one of the shop
+        if (!isset($_SERVER['HTTP_REFERER']) || $_SERVER['HTTP_REFERER'] !== Tools::secureReferrer($_SERVER['HTTP_REFERER'])) {
+            return null;
+        }
+
+        // Check if the previous page was a category page
```
Comment (Progi1984) [inline controllers/front/ProductController.php:1374]:
```suggestion
        // Check if the previous page was a category or a product page
```

File: controllers/front/ProductController.php [hunk 6]
```diff
@@ -1369,4 +1358,65 @@ public function setQuickViewMode(bool $enabled = true)
...
+
+        return null;
+    }
+
+    /**
+     * Given an id and a page link, extracts the category id
+     *
+     * @return int|null
+     */
+    private function extractObjectCategoryFromLink($id_object, $previousPageLink): ?int
```
Comment (Progi1984) [inline controllers/front/ProductController.php:1406]: Can you typehint all parameters of this method ?

File: controllers/front/ProductController.php [hunk 7]
```diff
@@ -1369,4 +1358,65 @@ public function setQuickViewMode(bool $enabled = true)
...
+    {
+        if (strncmp($this->context->link->getCategoryLink($id_object), $previousPageLink, strlen($this->context->link->getCategoryLink($id_object))) === 0) {
+            return (int) $id_object;
+        }
+
+        if (
+            isset($this->context->cookie->last_visited_category) &&
+            (int) $this->context->cookie->last_visited_category &&
```
Comment (Progi1984) [inline controllers/front/ProductController.php:1414]:
```suggestion
            !empty($this->context->cookie->last_visited_category) &&
```

**Review by Progi1984 — COMMENTED** [review]
@jbenezech Thank you for your contribution. Could you check my comments ?

**Comment by kpodemski** [comment]
Hello @jbenezech 

Sorry for bringing this topic so late, but...

I remember this issue; it was about `ps_categorytree`, not the core. I'm not sure if we should care about the latest visited category before visiting a product page 🤔 The problem with the module was that it was getting a category from the wrong source.

**Comment by matks** [comment]
Hello all, it looks like this PR is stale since April. I'll go ahead and close it. If someone is willing to continue this work please ping me we can reopen the PR. Or maybe open a new PR it will be easier 😉

**Pull request closed without merging**